src/Castle.Core/DynamicProxy/Generators/BaseClassProxyGenerator.cs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2004-2021 Castle Project - http://www.castleproject.org/
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+namespace Castle.DynamicProxy.Generators
+{
+	using System;
+	using System.Collections.Generic;
+	using System.Linq;
+	using System.Reflection;
+
+	using Castle.DynamicProxy.Contributors;
+	using Castle.DynamicProxy.Generators.Emitters.SimpleAST;
+	using Castle.DynamicProxy.Internal;
+
+	internal abstract class BaseClassProxyGenerator : BaseProxyGenerator
+	{
+		protected BaseClassProxyGenerator(ModuleScope scope, Type targetType, Type[] interfaces, ProxyGenerationOptions options)
+			: base(scope, targetType, interfaces, options)
+		{
+			EnsureDoesNotImplementIProxyTargetAccessor(targetType, "targetType");
+		}
+
+		protected abstract FieldReference TargetField { get; }
+
+		protected abstract ProxyInstanceContributor GetProxyInstanceContributor(List<MethodInfo> methodsToSkip);
+
+		protected abstract CompositeTypeContributor GetProxyTargetContributor(List<MethodInfo> methodsToSkip, INamingScope namingScope);
+
+		protected sealed override Type GenerateType(string name, INamingScope namingScope)
+		{
+			IEnumerable<ITypeContributor> contributors;
+			var allInterfaces = GetTypeImplementerMapping(out contributors, namingScope);
+
+			var model = new MetaType();
+			// Collect methods
+			foreach (var contributor in contributors)
+			{
+				contributor.CollectElementsToProxy(ProxyGenerationOptions.Hook, model);
+			}
+			ProxyGenerationOptions.Hook.MethodsInspected();
+
+			var emitter = BuildClassEmitter(name, targetType, allInterfaces);
+
+			CreateFields(emitter);
+			CreateTypeAttributes(emitter);
+
+			// Constructor
+			var cctor = GenerateStaticConstructor(emitter);
+
+			var constructorArguments = new List<FieldReference>();
+
+			if (TargetField is { } targetField)
```

Comment:
Is this basically checking if `TargetField != null`?

Reply:
Good catch... yes.

Btw. if you find more of those minor mess ups that may seem too small to fix individually, perhaps I'll just send one final PR that fixes them all together. Let me know which ones you want changed.

Reply:
No problem. I've added a few comments as I go. Happy for you to make one PR with a bunch of fixes.